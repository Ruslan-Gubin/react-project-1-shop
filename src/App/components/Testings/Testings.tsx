import React from 'react';

const Testings = () => {
  const [data, setData] = React.useState<null | {}>(null)
  const [toggle, setToggle] = React.useState(false)
  const [value, setValue] = React.useState('')

  const handlerToggle = () => setToggle(prev => !prev)

  React.useEffect(() => {
    setTimeout(() => {
      setData({})
    }, 100)
  },[])

  return (
    <div>
      <h1 data-testid='value-elem'>{value}</h1>
      {toggle && <div data-testid='toggle-elem'>toggle</div>}
      {data && <div style={{color: 'red'}} >data</div>}
      <h1>Hello world</h1>
      <button data-testid='toggle-btn' onClick={handlerToggle}>click me</button>
      <input onChange={(e) => setValue(e.target.value)} value={value} type="text" placeholder='input text..'/>
    </div>
  );
};

export  {Testings};