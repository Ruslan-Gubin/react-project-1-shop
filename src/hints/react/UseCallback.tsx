import React, { useCallback, useEffect, useState } from 'react';

const UseCallback = () => {
    const [message, useMessage] = useState('message')
   const[counter, setCounter] = useState(0) 

    const greeting = useCallback((text:string) => {
        console.log(text);
    },[])
    // useCallback обеспечиват ссылку на одну у туже функцию
// возвращает одну и туже функцию 

    useEffect(()=> {
        greeting(message)
    },[greeting, message]) // вызываем функцию которая вызывается при первом рендереге
    


    return (
        <div>
            <button onClick={() => setCounter(counter + 1)}>counter {counter}</button>
        </div>
    );
};

export default UseCallback;