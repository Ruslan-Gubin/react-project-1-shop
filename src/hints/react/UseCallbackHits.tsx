/**useCallback - возвращает мемонизированную версию функции которая изменяется только если изменяются 
 * значение одной из зависимостей. Мемонизированная означает что ссылка на эту функцию не будет 
 * менятся во время ререндеров, это нужно чтобы избежать ненужный рендеры дочерних компонентов которые пологаются 
 * на равенство ссылок
 */

// const CountComponent  = () => {

//     const [count, setCount] = useState(0)
    
//     const handlerCount = useCallback(() => {
//         setCount(count => count + 1)
//     } ,[])
    
//     return (
//         <div >
//             <ComponentHits handlerCount={handlerCount}/>
//         </div>
//     ) 
// }

/** оборачиваем функцию в useCallback и при каждом рендере ссылка на функцию не меняется 
/** ComponentHits оборачиваем в React.memo() и теперь  ComponentHits получает 
 *  функцию с одно и тойже ссылкой и больше не происходит ререндер дочернего компонента
*/