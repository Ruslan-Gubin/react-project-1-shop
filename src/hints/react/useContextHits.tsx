/**useContext - чтение значений контекста
 * контекст это данные которые видны на всем протяжении дерева компонентов, от родителя включая всех дочерних компонентов
 * контекст позволяет передавать данные по дереву без необходимости передавать пропсы на промежуточных уровнях
 * в целом контекст полезен однако он привязывает к себе компоненты и это усложняет использование повторно этих компонентов
 * 
 * 
 * 
 */
//@ts-ignore
// import { Component, createContext, useContext } from "react";

// /**создаем контекст передает дэфолтное состояние (необязательно)*/
// export const MyContext = createContext({default: false, theme: ''}) 

// /** оборачиваем родительский компонент провайдером */
// function AppContext() {
//   return (
//     <div>
//   <MyContext.Provider value={{default: false, theme: 'light'}>
//   <MyComponent />
//   <MyContext.Provider/>
//     </div>
//   )
// }

// /** достаем в дочернем компоненте  контекст*/
// const MyComponent = () => {
//   const {default, theme} = useContext(MyContext)
// }
