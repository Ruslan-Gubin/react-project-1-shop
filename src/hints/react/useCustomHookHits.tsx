/**
 * useCustomHook - это хук который пишет сам разработчик для того чтобы инкапсилировать логику
 * который работает с состоянием в отдельный хук, делается для того чтобы переиспользовать
 * логику в других компонентах
 */

// import { useCallback, useEffect, useRef, useState } from "react";

// const useResize = (ref) => {
//   const [width, setWidth] = useState(0)
//   const [height, setHeight] = useState(0)
//   const [squareSize, setSquareSize] = useState(0)

//   const handlerResize = useCallback(() => {
//     setWidth(ref.current.offsetWidth)
//     setHeight(ref.current.offsetHeight)
//     setSquareSize(Math.min(ref.current.offsetWidth,ref.current.offsetHeight ))
//   },[ref])

//   useEffect(() => {
//     handlerResize()

//     window.addEventListener('resize', handlerResize)
//     return () => {
//       window.removeEventListener('resize',handlerResize)
//     }
//   }, [handlerResize])

//   return { width, height, squareSize,}
// }

// const useToggleHits = () => {
//   const containerRef = useRef<HTMLDivElement>(null);
//   const { squareSize } = useResize(containerRef);

//   return (
//     <div className="root" ref={containerRef}>
//       <div
//         className="root-container"
//         style={{ width: `${squareSize}px`, height: `${squareSize}px` }}
//       ></div>
//     </div>
//   );
// };
