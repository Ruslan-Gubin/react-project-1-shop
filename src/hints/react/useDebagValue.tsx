/** useDebagValue - для косоли разработчика react devtools видеть некоторые значение
 * которые будет удобно воспринимать при дебаге этого хука 
 */

// import { useDebugValue } from "react"

// const CustomComponent = () => {
// const obj = {
//   name: 'Ruslan',
//   age: 30
// }

// useDebugValue(
//   obj,
//   (obj:{name:string, age:number}) => `w: ${obj.name}; a: ${obj.age}` 
// )

//   return obj
// }