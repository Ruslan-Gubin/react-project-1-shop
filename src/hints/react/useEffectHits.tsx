/**useEffect - применяется сайд эфектов, тоесть побочных функций которые должны выполнятся
 * на изменение пропсов либо состояние , это може запрос данных с сервера установка таймеров интервалов, 
 * инициализация подписок и так далее, вторым аргументом содержит массив пропсов или стейтов которые 
 * нужно отслеживать, когда их значение будут меняться то функция эфекта будетвызвана заново
 * Если массив зависимости пустой Функция колбек из useEffect отрабатывает только после самого первого рендера 
 * useEffect - может возвращать другую  функцию , эта функция возвращается при анмаунте компонента,
 * либо при кажном изменении пропсов или стейта содержащее в массиве зависимости, и эта цункция 
 * предназначена для чтого чтобы очищать таймеры таймауты либо очищать подписки eventListener и т.д
 * Для того чтобы избежать утечек памяти или ошибок связанных с тем подписка с которой было взаимодействие
 * например с удаленным компонентом она остается 
 */

React.useEffect(() => {
  /**тело колбек ценкции */

  return () => {
  /**возвращаемая функция, отписка от событий */
  } 

},[/**массив зависимости */])