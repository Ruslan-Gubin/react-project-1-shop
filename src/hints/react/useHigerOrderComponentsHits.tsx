/**Higer-OrderComponents - компонент высшего порядка (hok)
 * функция которая принимает компонент в качестве агрумента и возвращает  другой компонент
 * называется компонентом высшего порядка, он применяется для повторного использования логики
 * тоесть он помогает следовать принципу dry (dont repeat Youself) не повторяйся
 * не стоит изиенять компонент внутри компонента высшего порядка, например не нужно переопределять
 * методы жизненого цикла. 
  */