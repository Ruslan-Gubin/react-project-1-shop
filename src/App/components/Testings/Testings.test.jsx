import '@testing-library/jest-dom';
import {render, screen, fireEvent} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import {Testings} from './Testings';

describe('Testings component', () => {

  
  it('render Testings components ', () => {
    render(<Testings />)
    const hellowWorldElem = screen.getByText(/Hello world/i);
    const btn = screen.getByRole('button')
    const input = screen.getByPlaceholderText(/input text/i)
    expect(btn).toBeInTheDocument()
    expect(input).toBeInTheDocument()
    expect(hellowWorldElem).toBeInTheDocument()
    expect(input).toMatchSnapshot()
  })
  it('Testings components is not Hello world2', () => {
    render(<Testings />)
    expect(screen.queryByText(/world2/i)).toBeNull()
  })
  it('Testings async element ', async () => {
    render(<Testings />)
    const elem = await screen.findByText(/data/i)
    expect(elem).toBeInTheDocument()
    expect(elem).toHaveStyle({color: 'red'})
  })
  it('Toggle event ', async () => {
    render(<Testings />)
   const btn = screen.getByTestId('toggle-btn')
   expect(screen.queryByTestId('toggle-elem')).toBeNull()
   fireEvent.click(btn)
   expect(screen.queryByTestId('toggle-elem')).toBeInTheDocument()
   fireEvent.click(btn)
   expect(screen.queryByTestId('toggle-elem')).toBeNull()
  })
  it('Input event ', async () => {
    render(<Testings />)
    const input = screen.getByPlaceholderText(/input text../i)
   expect(screen.queryByTestId('value-elem')).toContainHTML('')
   fireEvent.input(input, {
    target: {value: '123123'}
   })
   expect(screen.queryByTestId('value-elem')).toContainHTML('123123')
   expect(screen.queryByTestId('value-elem')).not.toBeNull()
  })
  test('Input User event ', async () => {
    render(<Testings />)
    const input = screen.getByPlaceholderText(/input text../i)
   expect(screen.getByTestId('value-elem')).toContainHTML('')
 await  userEvent.type(input, '123123')
   expect(screen.queryByTestId('value-elem')).toContainHTML('123123')
  })
  
})