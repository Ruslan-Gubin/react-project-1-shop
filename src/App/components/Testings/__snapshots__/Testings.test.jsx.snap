// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testings component render Testings components  1`] = `
<input
  placeholder="input text.."
  type="text"
  value=""
/>
`;
